### PR TITLE
Labels support

### PR DESCRIPTION
**Pull request by danni-m**

### CODE REVIEW

File: src/indexer.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+#include "consts.h"
+#include "indexer.h"
+
+int parseLabel(RedisModuleCtx *ctx, RedisModuleString *label, Label *retLabel, const char *separator) {
+    char *token;
+    char *iter_ptr;
+    size_t _s;
+    const char *labelRaw = RedisModule_StringPtrLen(label, &_s);
+    char *labelstr = RedisModule_PoolAlloc(ctx, _s + 1);
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: why is this needed? I see from the code that RedisModule_StringPtrLen returns a null terminated string

File: src/indexer.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    }
+    return TSDB_OK;
+}
+
+void IndexOperation(RedisModuleCtx *ctx, const char *op, RedisModuleString *ts_key, Label *labels, size_t labels_count) {
+    RedisModuleCallReply *reply;
+
+    for (int i=0; i<labels_count; i++) {
+        size_t _s;
+        RedisModuleString *sk = RedisModule_CreateStringPrintf(ctx, "__index_%s=%s",
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: what does sk mean?

File: src/indexer.c [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    IndexOperation(ctx, "SADD", ts_key, labels, labels_count);
+}
+
+void RemoveIndexedMetric(RedisModuleCtx *ctx, RedisModuleString *ts_key, Label *labels, size_t labels_count) {
+    IndexOperation(ctx, "SREM", ts_key, labels, labels_count);
+}
+
+
+// Go over the results dictionary (dictEq) and remove keys that didn't match on all equal matchers
+int _remove_not_intersected(int eqCount, RedisModuleDict *dictEq, char **lastKey, size_t *lastKeySize) {
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: AFAIK this is not a naming convention in C

File: src/consts.h [hunk 4]
```diff
@@ -6,8 +6,8 @@
 #define TRUE 1
 #define FALSE 0
 
-#define timestamp_t int32_t
-#define api_timestamp_t int32_t
+#define timestamp_t u_int32_t
```
Comment (yaeltzirulnikov) [inline src/consts.h]: why is it still 32 and not 64?

File: src/indexer.c [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        if (token == NULL) {
+            return TSDB_ERROR;
+        }
+        if (i == 0) {
+            retLabel->key = RedisModule_CreateString(NULL, token, strlen(token));
+        } else {
+            retLabel->value = RedisModule_CreateString(NULL, token, strlen(token));
+        }
+        token = strtok_r (NULL, "=", &iter_ptr);
```
[the remark this answers is not in the document]
Comment (danni-m) [reply src/indexer.c]: @yaeltzirulnikov yes, good catch. it might have worked because the initial strtok_r was with `separator`

File: src/indexer.c (hunk 1 above)
Comment (danni-m) [reply src/indexer.c]: @yaeltzirulnikov strtok_r manipulates the string its recommend to use it on a copied string.

File: src/indexer.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (danni-m) [reply src/indexer.c]: fixed.

File: src/indexer.c (hunk 2 above)
Comment (danni-m) [reply src/indexer.c]: renamed

File: src/indexer.c [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    RedisModuleCallReply *reply;
+
+    for (int i=0; i<labels_count; i++) {
+        size_t _s;
+        RedisModuleString *indexed_key_value = RedisModule_CreateStringPrintf(ctx, "__index_%s=%s",
+                RedisModule_StringPtrLen(labels[i].key, &_s),
+                RedisModule_StringPtrLen(labels[i].value, &_s));
+        RedisModuleString *indexed_key = RedisModule_CreateStringPrintf(ctx, "__index_%s",
+                RedisModule_StringPtrLen(labels[i].key, &_s),
+                RedisModule_StringPtrLen(labels[i].value, &_s));
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: this line should be removed, only key is expected

Comment (yaeltzirulnikov) [reply src/indexer.c]: also, let's not call  RedisModule_StringPtrLen(labels[i].key, &_s) twice

File: src/indexer.c [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+void RemoveIndexedMetric(RedisModuleCtx *ctx, RedisModuleString *ts_key, Label *labels, size_t labels_count) {
+    IndexOperation(ctx, "SREM", ts_key, labels, labels_count);
+}
+
+int _intersect(RedisModuleDict *left, RedisModuleDict *right, char **lastKey, size_t *lastKeySize) {
+    RedisModuleDictIter *iter;
+    if (*lastKey == NULL) {
+        iter = RedisModule_DictIteratorStartC(left, "^", NULL, 0);
+    } else {
+        iter = RedisModule_DictIteratorStartC(left, "^", lastKey, *lastKeySize);
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: I'm not sure what this line is trying to do, but RedisModule_DictIteratorStartC with "^" ignores the given key

File: src/indexer.c [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        } else  if (predicate->type == NCONTAINS) {
+            while (_difference(prevResults, localResult, &lastKey, &lastKeySize) != 0) {}
+        } else if (predicate->type == NEQ){
+            while (_difference(prevResults, localResult, &lastKey, &lastKeySize) != 0) {}
+        }
+        return prevResults;
+    } else if (predicate->type == EQ) {
+        return localResult;
+    } else {
+        return prevResults;
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: if you try to find all keys that don't match a label , and prevResults is NULL, this will return NULL instead of all the keys
